elephant/causality/granger.py

```diff
@@ -596,3 +776,163 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
+                accounted for by the directional causalities (e.g. shared input
+                to X and Y).
+
+            total_interdependence : np.ndarray
+                The sum of the former three metrics. It measures the dependence
+                of X and Y. If the total interdependence is positive, X and Y
+                are not independent.
+    """
+    if isinstance(signal_i, neo.core.AnalogSignal) and \
+            isinstance(signal_j, neo.core.AnalogSignal):
+        signals = signal_i.merge(signal_j)
+    elif isinstance(signal_i, np.ndarray) and isinstance(signal_j, np.ndarray):
+        signals = np.vstack([signal_i, signal_j])
+
+    # Calculate cross spectrum for signals
+    freqs, S = segmented_multitaper_cross_spectrum(
+        signals=signals, n_segments=n_segments, len_segment=len_segment,
+        frequency_resolution=frequency_resolution, overlap=overlap, fs=fs,
+        nw=nw, num_tapers=num_tapers, peak_resolution=peak_resolution,
+        return_onesided=False)
+
+    # Remove units attached by the multitaper_cross_spectrum
+    if isinstance(S, pq.Quantity):
+        S = S.magnitude
+
+    # Transpose cross spectrum due to different conventions used in
+    # segemented_multitaper_cross_spectrum and the calculations of spectral
+    # Granger causality
+    S = np.transpose(S, axes=(1, 0, 2))
+
+    # Decompose cross spectrum into covariance and transfer function
+    C, H = _spectral_factorization(S, num_iterations=num_iterations)
+
+    # Take positive frequencies
+    mask = (freqs >= 0)
+    freqs = freqs[mask]
+
+    S = np.transpose(S, axes=(2, 0, 1))[mask]
+    H = H[mask]
+
+    # Calculate spectral Granger Causality.
+    # Formulae follow Wen et al., 2012, Phil Trans R Soc
+    H_tilde_xx = H[:, 0, 0] + C[0, 1]/C[0, 0]*H[:, 0, 1]
+    H_tilde_yy = H[:, 1, 1] + C[0, 1]/C[1, 1]*H[:, 1, 0]
+
+    directional_causality_y_x = np.log(S[:, 0, 0].real /
+                                       (H_tilde_xx
+                                        * C[0, 0]
+                                        * H_tilde_xx.conj()).real)
+
+    directional_causality_x_y = np.log(S[:, 1, 1].real /
+                                       (H_tilde_yy
+                                        * C[1, 1]
+                                        * H_tilde_yy.conj()).real)
+
+    instantaneous_causality = np.log(
+        (H_tilde_xx * C[0, 0] * H_tilde_xx.conj()).real
+        * (H_tilde_yy * C[1, 1] * H_tilde_yy.conj()).real)
+    instantaneous_causality -= np.linalg.slogdet(S)[1]
+
+    total_interdependence = (directional_causality_x_y
+                             + directional_causality_y_x
+                             + instantaneous_causality)
+
+    spectral_causality = Causality(
+        directional_causality_x_y=directional_causality_x_y,
+        directional_causality_y_x=directional_causality_y_x,
+        instantaneous_causality=instantaneous_causality,
+        total_interdependence=total_interdependence)
```

Comment:
Is there a particular reason why to make this into such an object? I have not seen such an output anywhere else, it might confuse some people when using it

Reply:
As discussed in person, agree to keep this